### PR TITLE
[Merged by Bors] - chore: use etaExperiment rather than hacking with instances

### PR DESCRIPTION
**Pull request by kim-em**
This is to fix timeouts in https://github.com/leanprover-community/mathlib4/pull/3552.

See discussion at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/!4.233552.20.28LinearAlgebra.2EMatrix.2EToLin.29.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This is to fix timeouts in https://github.com/leanprover-community/mathlib4/pull/3552.

**Comment by semorrison** [comment]
I have not exhaustively removed evil instances hackery in favour of `etaExperiment` --- just the hackery that was implicated in the timeouts in https://github.com/leanprover-community/mathlib4/pull/3552. If this is approved I can optionally do the remainder of the cleanup in a subsequent PR.

**Comment by jcommelin** [comment]
CI is unhappy...

File: Mathlib/Data/Polynomial/Module.lean [hunk 1]
```diff
@@ -196,7 +197,11 @@ theorem smul_apply (f : R[X]) (g : PolynomialModule R M) (n : ℕ) :
     exacts [rfl, (zero_smul R _).symm]
 #align polynomial_module.smul_apply PolynomialModule.smul_apply
 
+-- Porting note: typeclass inference goes haywire in the next declaration.
+-- One way to prune the bad branches is:
+attribute [-instance] IsDomain.toCancelCommMonoidWithZero
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Module.lean:202]: This looks like more of the thing we're trying to remove in this PR; is it worth mentioning `etaExperiment` or `2074` in the comment for later discoverability?

Also, should this be

```suggestion
-- Porting note: typeclass inference goes haywire in the next declaration.
-- One way to prune the bad branches is:
attribute [-instance] IsDomain.toCancelCommMonoidWithZero in
```
to limit it to the one declaration?

File: Mathlib/RingTheory/QuotientNoetherian.lean [hunk 2]
```diff
@@ -15,6 +15,9 @@ import Mathlib.RingTheory.QuotientNilpotent
 # Noetherian quotient rings and quotient modules
 -/
 
+-- Porting note: we keep this instance local to avoid downstream effects.
+-- I haven't been able to work out how to omit it or inline it into the the construction below.
+local instance {R : Type _} [CommRing R] : Module R R := eta_experiment% inferInstance
```
Comment (eric-wieser) [inline Mathlib/RingTheory/QuotientNoetherian.lean:20]: This porting note could do with a reference to 2075 as well

**Review by eric-wieser — APPROVED** [review]
bors d+

Let's make sure to keep the new porting notes easy to find

File: lakefile.lean [hunk 3]
```diff
@@ -4,6 +4,7 @@ open Lake DSL
 
 def moreLeanArgs := #[
   "-DwarningAsError=true",
+  "-DsynthInstance.etaExperiment=true",
```
Comment (eric-wieser) [inline lakefile.lean:7]: Did you intend to commit this?

Comment (semorrison) [reply lakefile.lean:7]: oops! I was just experimenting with that locally now... :-) Will revert.

**Comment by semorrison** [comment]
!bench

**Comment by semorrison** [comment]
bors merge

**Comment by jcommelin** [comment]
Do those benchmark results mean that `Mathlib.LinearAlgebra.Span` became ~50% slower?

**Pull request closed without merging**